synapse/config/saml2_config.py

```diff
@@ -75,14 +69,58 @@ def read_config(self, config, **kwargs):
 
         self.saml2_enabled = True
 
-        self.saml2_mxid_source_attribute = saml2_config.get(
-            "mxid_source_attribute", "uid"
-        )
-
         self.saml2_grandfathered_mxid_source_attribute = saml2_config.get(
             "grandfathered_mxid_source_attribute", "uid"
         )
 
+        user_mapping_provider_dict = saml2_config.get("user_mapping_provider")
+        if user_mapping_provider_dict is None:
+            user_mapping_provider_dict = {}
+
+        # Load configured module class and config
+        default_mapping_provider = (
+            "synapse.handlers.saml_handler.DefaultSamlMappingProvider"
+        )
+        user_mapping_provider_module = user_mapping_provider_dict.get(
+            "module", default_mapping_provider,
+        )
+        user_mapping_provider_config = user_mapping_provider_dict.get("config", {})
+
+        if user_mapping_provider_module == default_mapping_provider:
+            # Handle deprecated options in default module config
+
+            # If mxid_source_attribute is defined in the deprecated location, use
+            # that instead for backwards compatibility
+            old_mxid_source_attribute = saml2_config.get("mxid_source_attribute")
+            if old_mxid_source_attribute:
+                user_mapping_provider_config[
+                    "mxid_source_attribute"
+                ] = old_mxid_source_attribute
+
+                # Provide a deprecation warning
+                self.using_old_mxid_source_attribute = True
+            else:
+                # If the option doesn't exist in saml2_config, and it's not set under
+                # user_mapping_provider_config, set it to a default value
+                user_mapping_provider_config.setdefault("mxid_source_attribute", "uid")
```

Comment:
the default value should be up to the mapping provider to sort out.

Reply:
This would be nice, though we also need to take the grandfathered value into account. It's not part of the module's configuration dict, so I'm not sure how to get it to the module.

Reply:
which grandfathered value?

Reply:
Ah, there's `SAML2Config.saml2_grandfathered_mxid_source_attribute`, but it's not actually used outside of the module. No need to worry about it here.